pkg/crmutil/controller-data.go

```diff
@@ -200,12 +200,40 @@ type NeedsUpdate struct {
 	AppInstRuntime bool
 }
 
+func (cd *CRMHandler) clusterInstDNSChanged(ctx context.Context, pf platform.Platform, oldFQDN string, new *edgeproto.ClusterInst, sender edgeproto.ClusterInstInfoSender) (reterr error) {
+	var err error
+
+	log.SpanLog(ctx, log.DebugLevelInfra, "ClusterInstChange for DNS", "key", new.Key, "old dns", oldFQDN, "new dns", new.Fqdn)
+
+	updateClusterCacheCallback := sender.SendStatusIgnoreErr
+	// reset status messages
+	err = sender.SendState(edgeproto.TrackedState_UPDATING, edgeproto.WithSenderResetStatus())
+	if err != nil {
+		return err
+	}
+	err = pf.ChangeClusterInstDNS(ctx, new, oldFQDN, updateClusterCacheCallback)
+	if err != nil {
+		err := fmt.Errorf("update failed: %s", err)
+		sender.SendState(edgeproto.TrackedState_UPDATE_ERROR, edgeproto.WithStateError(err))
+		return err
+	}
+	log.SpanLog(ctx, log.DebugLevelInfra, "ClusterInstChange for DNS done", "key", new.Key, "old dns", oldFQDN, "new dns", new.Fqdn)
+	sender.SendState(edgeproto.TrackedState_READY)
+	return nil
+}
+
 func (cd *CRMHandler) ClusterInstChanged(ctx context.Context, target *edgeproto.CloudletKey, new *edgeproto.ClusterInst, sender edgeproto.ClusterInstInfoSender) (nu NeedsUpdate, reterr error) {
 	var err error
 
 	fmap := edgeproto.MakeFieldMap(new.Fields)
 
-	if !fmap.Has(edgeproto.ClusterInstFieldState) {
+	dnsUpdate := false
+	oldFQDN, ok := new.Annotations[cloudcommon.AnnotationPreviousDNSName]
+	if fmap.Has(edgeproto.ClusterInstFieldFqdn) && ok && oldFQDN != new.Fqdn {
+		dnsUpdate = true
+	}
```

Comment:
You should just check if the field flag is set. The oldFQDN in the annotation is always going to be different than the current FQDN.

Reply:
that's fair - `oldFQDN != new.Fqdn` is a noop check. I'll keep the check for the existence of annotation though.